File: web/src/owner-handoff.mjs
import { cloudAPI, cloudURL } from './managed-clouds.mjs';
const uuid = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
export const handoffAcceptPath = '/brand-cloud-owner-transfer/accept';
export function handoffRoute(path) {
  if (path === handoffAcceptPath) return { accept: true };
  const m = path.match(/^\/console\/clouds\/([^/]+)\/owner-transfer\/([^/]+)\/?$/);
  return m && uuid.test(m[1]) && uuid.test(m[2]) ? { cloudId: m[1], transferId: m[2] } : null;
}
export function handoffURL(cloud, transfer) { if (!uuid.test(transfer)) throw new Error('Invalid transfer'); return `${cloudURL(cloud)}/owner-transfer/${transfer}`; }
export function handoffAPI(cloud, transfer = '', action = '') {
  if (transfer && !uuid.test(transfer)) throw new Error('Invalid transfer');
  if (action && (!transfer || !['preview', 'confirm', 'cancel'].includes(action))) throw new Error('Invalid action');
  return `${cloudAPI(cloud)}/owner-transfer${transfer ? `/${transfer}` : ''}${action ? `/${action}` : ''}`;
}
export function safeSnapshot(s) {
  return Boolean(s && Number.isSafeInteger(s.ownership_version) && s.ownership_version >= 1 && Number.isSafeInteger(s.billing_snapshot_version) && s.billing_snapshot_version >= 2 && Number.isSafeInteger(s.balance_minor) && s.balance_minor >= 0 && s.currency === 'TWD');
}
export function snapshotKey(v) { return safeSnapshot(v?.balance_snapshot) ? JSON.stringify([v.brand_cloud_id, v.id, v.balance_snapshot.ownership_version, v.balance_snapshot.billing_snapshot_version, v.balance_snapshot.balance_minor, v.balance_snapshot.currency]) : ''; }
export function handoffConfirmable(v, actor) {
  return Boolean(v?.phase === 'awaiting_balance_confirmation' && v.operation_phase === 'preparing' && v.has_settled_snapshot && safeSnapshot(v.balance_snapshot) && v.balance_snapshot.ownership_version === v.ownership_version && v.blockers?.length === 0 && (actor === v.source_user_id ? v.source_confirmed === false : actor === v.target_user_id && v.target_confirmed === false));
}
export function handoffCancelable(v, actor) { return Boolean(actor && v?.source_user_id === actor && ['pending', 'accepted'].includes(v.status) && !['canceling', 'canceled', 'finalizing', 'succeeded'].includes(v.operation_phase) && !['succeeded', 'finalizing', 'expired', 'canceled'].includes(v.phase)); }
export function handoffComplete(v) { return v?.phase === 'succeeded' && v.operation_phase === 'succeeded' && v.operation?.state === 'succeeded'; }
export function handoffTitle(v) {
  if (handoffComplete(v)) return 'Ownership handoff completed';
  if (v?.operation_phase === 'canceling') return 'Cancellation awaiting hold release';
  if (v?.phase === 'finalizing') return 'Owner changed; finalization in progress';
  if (v?.phase === 'canceled') return 'Handoff canceled; holds released';
  if (v?.phase === 'expired') return 'Ownership invitation expired';
  return 'Handoff in progress';
}
export function handoffError(err) {
  if (err?.status === 401) return 'Your session expired. Sign in again.';
  if ([403, 404].includes(err?.status)) return 'This account cannot access this handoff, or the invitation has expired.';
  if (err?.status === 409) return 'The handoff is blocked or its snapshot changed. Review status and request a fresh settled preview. Negative balance, unsettled usage, pending payments, quota or ownership changes can block transfer.';
  if ([400, 422].includes(err?.status)) return 'Check the invitee and exact nonnegative balance snapshot.';
  return 'Handoff status is unavailable. Do not assume ownership changed or retry with a different confirmation amount.';
}

File: internal/accountclient/owner_handoff.go
package accountclient

import (
	"context"
	"fmt"
	"net/http"
	"strings"
)

type HandoffSnapshot struct {
	OwnershipVersion       int64  `json:"ownership_version"`
	BillingSnapshotVersion int64  `json:"billing_snapshot_version"`
	BalanceMinor           int64  `json:"balance_minor"`
	Currency               string `json:"currency"`
}
type OwnerHandoff struct {
	ID                 string                 `json:"id"`
	CloudID            string                 `json:"brand_cloud_id"`
	SourceUserID       string                 `json:"source_user_id"`
	RequestedByUserID  string                 `json:"requested_by_user_id"`
	TargetUserID       string                 `json:"target_user_id"`
	TargetEmail        string                 `json:"target_email,omitempty"`
	Status             string                 `json:"status"`
	Phase              string                 `json:"phase"`
	OperationPhase     string                 `json:"operation_phase,omitempty"`
	HasSettledSnapshot bool                   `json:"has_settled_snapshot"`
	BalanceSnapshot    *HandoffSnapshot       `json:"balance_snapshot,omitempty"`
	SourceConfirmed    *bool                  `json:"source_confirmed,omitempty"`
	TargetConfirmed    *bool                  `json:"target_confirmed,omitempty"`
	OwnershipVersion   int64                  `json:"ownership_version,omitempty"`
	Operation          *ManagedCloudOperation `json:"operation,omitempty"`
	Blockers           []CloudBlocker         `json:"blockers"`
	ExpiresAt          string                 `json:"expires_at"`
}

// Participant endpoints must not use current cloud membership: the recipient
// has none before commit and the former owner loses theirs after commit.
func (c *Client) OwnerHandoff(ctx context.Context, token, subject, method, path, cloud, transfer, key string, body any) (OwnerHandoff, error) {
	var result struct {
		Transfer OwnerHandoff `json:"owner_transfer"`
	}
	if err := c.doJSONWithIdempotency(ctx, method, path, token, key, body, &result); err != nil {
		return OwnerHandoff{}, err
	}
	v := result.Transfer
	if v.Blockers == nil {
		return v, fmt.Errorf("missing handoff blocker evidence")
	}
	if !managedOperationID.MatchString(v.ID) || !managedOperationID.MatchString(v.CloudID) || (cloud != "" && v.CloudID != cloud) || (transfer != "" && v.ID != transfer) || v.SourceUserID == "" || v.SourceUserID != v.RequestedByUserID || v.TargetUserID == "" || v.TargetUserID == v.SourceUserID || (subject != v.SourceUserID && subject != v.TargetUserID) {
		return v, fmt.Errorf("invalid handoff participant or scope")
	}
	if strings.HasSuffix(path, "/accept") && subject != v.TargetUserID {
		return v, fmt.Errorf("handoff recipient mismatch")
	}
	if method == http.MethodPost && transfer == "" && cloud != "" && subject != v.SourceUserID {
		return v, fmt.Errorf("handoff source mismatch")
	}
	switch v.Phase {
	case "requested", "awaiting_acceptance", "preparing", "awaiting_balance_confirmation", "committing", "finalizing", "succeeded", "blocked", "canceled", "expired":
	default:
		return v, fmt.Errorf("missing or unknown handoff phase")
	}
	if v.HasSettledSnapshot != (v.BalanceSnapshot != nil) || v.HasSettledSnapshot != (v.SourceConfirmed != nil && v.TargetConfirmed != nil) || (!v.HasSettledSnapshot && (v.SourceConfirmed != nil || v.TargetConfirmed != nil)) {
		return v, fmt.Errorf("incomplete handoff snapshot")
	}
	if snap := v.BalanceSnapshot; snap != nil && (snap.OwnershipVersion < 1 || snap.OwnershipVersion != v.OwnershipVersion || snap.BillingSnapshotVersion < 2 || snap.BalanceMinor < 0 || snap.Currency != "TWD") {
		return v, fmt.Errorf("invalid handoff snapshot")
	}
	if op := v.Operation; op != nil && (op.ID != v.ID || op.CloudID != v.CloudID || op.Type != "owner_transfer") {
		return v, fmt.Errorf("invalid handoff operation")
	}
	if v.Phase == "awaiting_balance_confirmation" && (v.BalanceSnapshot == nil || len(v.Blockers) != 0 || v.Operation == nil || v.OperationPhase != "preparing") {
		return v, fmt.Errorf("unready confirmation response")
	}
	if (v.Phase == "finalizing" || v.Phase == "succeeded") && (v.Operation == nil || v.OperationPhase != v.Phase || v.Status != "accepted" || v.BalanceSnapshot == nil || !*v.SourceConfirmed || !*v.TargetConfirmed) {
		return v, fmt.Errorf("unproven ownership commit")
	}
	if v.Phase == "succeeded" && v.Operation.State != "succeeded" {
		return v, fmt.Errorf("incomplete finalization")
	}
	return v, nil
}
